### PR TITLE
[IR] Support local tensor

### PR DESCRIPTION
**Pull request by squarefk**
## IR
Python:
```
a = ti.Vector([1., 2., 3.])

print(a[?])
```

Initial IR:
```
<[TensorType> $1 = alloca @tmp     // FrontendAllocaStmt(IdExpression, shape, element_type)
@tmp[0] = elements[0]              // FrontendAssignStmt(LocalTensorElementExpression(IdExpression, indices), elements)
@tmp[1] = elements[1]
@tmp[2] = elements[2]
$? = eval @tmp[?]                  // FrontendEvalStmt
print $?
```

Lower AST (eliminate IdExpression):
```
<TensorType> $1 = alloca @tmp      // AllocaStmt
<Type*> $2 = shift ptr [$1 + $?]   // ShiftGlobalPtrStmt
<Type> $3 = local store [$2 <- $?] // LocalStore

<Type> $4 = local load [$2]        // LocalLoad
```

Offloaded:
AllocaStmt may be converted into GlobalTemporaryStmt or just keep AllocaStmt

## Issues
Currently, there are three issues:
- <del>Offload pass only deal with ConstStmt type for offset in PtrOffsetStmt. This will introduce some code restrictions but should be easy to resolve. </del>. Done with another PR.
- During make_local_block pass, Alloca TensorType need LocalStore to initialize which introduces extra difficulty. Hence, it is disabled.
- Loop index keep using individual variable instead of TensorType for now. We may want to change this with another PR. At least the following aspects should be considered.
  2. for I in ti.grouped(ti.ndrange()): need to make sure Alloca TensorType inside loop body.
  3. Loop indices are looked up by Alloca id in some places. Extra work is needed to extract different dim indices from the same Alloca TensorType

### CODE REVIEW

File: taichi/transforms/offload.cpp [hunk 1]
```diff
@@ -426,18 +441,90 @@ class FixCrossOffloadReferences : public BasicStmtVisitor {
...
+    replacement.push_back<PtrOffsetStmt>(ptr, stmt->offset);
+
+    stmt->parent->replace_with(stmt, std::move(replacement));
+    throw IRModified();
+  }
+
   // Replace local LD/ST with global LD/ST
   void visit(LocalLoadStmt *stmt) override {
     TI_ASSERT(stmt->width() == 1);
+    // TensorType Alloca
```
Comment (k-ye) [inline taichi/transforms/offload.cpp:504]: Could you create a function for these three Stmts?

File: taichi/ir/statements.h [hunk 2]
```diff
@@ -25,6 +25,11 @@ class AllocaStmt : public Stmt {
     TI_STMT_REG_FIELDS;
   }
 
+  AllocaStmt(std::vector<int> shape, DataType type) {
```
Comment (k-ye) [inline taichi/ir/statements.h:28]:
```suggestion
  AllocaStmt(const std::vector<int> &shape, DataType type) {
```

File: taichi/ir/statements.h [hunk 3]
```diff
@@ -311,9 +316,23 @@ class PtrOffsetStmt : public Stmt {
...
+    return origin->is<AllocaStmt>() || origin->is<GlobalTemporaryStmt>();
+  }
+
+  bool is_unlowered_global_ptr() const {
+    return origin->is<GlobalPtrStmt>();
+  }
+
+  bool is_lowered_global_ptr() const {
+    // TODO: Check this is the only case.
+    return origin->is<GetChStmt>();
```
Comment (k-ye) [inline taichi/ir/statements.h:335]: Feels like a pretty restrictive assumption. Let's just use `!is_unlowered_global_ptr()` for now.

File: python/taichi/lang/impl.py [hunk 4]
```diff
@@ -16,12 +16,21 @@
 import taichi as ti
 
 
+@taichi_scope
+def expr_init_local_tensor(shape, element_type, elements):
+    ti.current_cfg().make_block_local = False
```
Comment (k-ye) [inline python/taichi/lang/impl.py:21]: Which test(s) does this break?

File: taichi/python/export_lang.cpp [hunk 5]
```diff
@@ -649,6 +649,29 @@ void export_lang(py::module &m) {
...
+      int linearized_index = i;
+      for (int d = (int)shape.size() - 1; d >= 0; --d) {
+        reversed_indices.push_back(
+            Expr::make<ConstExpression, int32>(linearized_index % shape[d]));
+        linearized_index /= shape[d];
+      }
+      for (int d = 0; d < (int)shape.size(); ++d)
+        indices.push_back(reversed_indices[(int)shape.size() - 1 - d]);
+      current_ast_builder().insert(std::make_unique<FrontendAssignStmt>(
+          Expr::make<LocalTensorElementExpression>(var, indices),
```
Comment (k-ye) [inline taichi/python/export_lang.cpp:670]: Is it possible to unify `LocalTensorElementExpression` and `GlobalTensorElementExpression`?

File: python/taichi/lang/matrix.py [hunk 6]
```diff
@@ -256,7 +292,11 @@ def linearize_entry_id(self, *args):
     def __call__(self, *args, **kwargs):
         _taichi_skip_traceback = 1
         assert kwargs == {}
-        return self.entries[self.linearize_entry_id(*args)]
+        if self.local_tensor_proxy is not None:
```
Comment (strongoier) [inline python/taichi/lang/matrix.py:295]: I believe this branch is redundant - this path is already prohibited in `subscript()`.

File: taichi/analysis/alias_analysis.cpp [hunk 7]
```diff
@@ -13,8 +13,39 @@ AliasResult alias_analysis(Stmt *var1, Stmt *var2) {
...
+      return var->cast<PtrOffsetStmt>()->origin;
+    } else {
+      return (Stmt *)nullptr;
+    }
+  };
+  Stmt *origin1 = retrieve_local(var1);
+  Stmt *origin2 = retrieve_local(var2);
+  if (origin1 != nullptr && origin2 != nullptr) {
+    if (origin1 == origin2)
+      return AliasResult::uncertain;
```
[the remark this answers is not in the document]
Comment (squarefk) [reply taichi/analysis/alias_analysis.cpp:33]: I have an inline function called ```retrieve_local```. When ```origin1==origin2``` does satisfy and L12 doesn't, it means that at least one of them involves pointer operation with PtrOffset.

File: taichi/ir/expr.cpp [hunk 8]
```diff
@@ -168,13 +168,16 @@ Expr load_if_ptr(const Expr &ptr) {
...
+    return Expr::make<LocalLoadExpression>(ptr);
   } else
     return ptr;
 }
 
+// TODO: remove this
 Expr load(const Expr &ptr) {
   TI_ASSERT(ptr.is<GlobalPtrExpression>() ||
-            ptr.is<GlobalTensorElementExpression>());
+            ptr.is<GlobalTensorElementExpression>())
```
[the remark this answers is not in the document]
Comment (squarefk) [reply taichi/ir/expr.cpp]: https://github.com/taichi-dev/taichi/pull/2637/commits/28656cae692a4737f1700e0d7f342d6085fae061

File: taichi/transforms/offload.cpp (hunk 1 above)
Comment (squarefk) [reply taichi/transforms/offload.cpp:504]: This has been done in another PR for simplifying offload pass and supporting the first issue.

File: taichi/ir/statements.h (hunk 3 above)
Comment (squarefk) [reply taichi/ir/statements.h:335]: Do you mean ```!is_local_ptr() && !is_unlowered_global_ptr()```?

File: taichi/python/export_lang.cpp (hunk 5 above)
Comment (squarefk) [reply taichi/python/export_lang.cpp:670]: Sounds great. Shall we open another PR for this?

File: python/taichi/lang/impl.py (hunk 4 above)
Comment (squarefk) [reply python/taichi/lang/impl.py:21]: ======================================== short test summary info ========================================
FAILED tests/python/test_bls.py::test_gather_3d - RuntimeError: [insert_scratch_pad.cpp:initialize_scr...
FAILED tests/python/test_bls.py::test_gather_1d_trivial - RuntimeError: [insert_scratch_pad.cpp:initia...
FAILED tests/python/test_bls.py::test_scatter_1d_trivial - RuntimeError: [insert_scratch_pad.cpp:initi...
FAILED tests/python/test_bls.py::test_gather_2d_nonsquare - RuntimeError: [insert_scratch_pad.cpp:init...
FAILED tests/python/test_bls.py::test_gather_1d - RuntimeError: [insert_scratch_pad.cpp:initialize_scr...
FAILED tests/python/test_bls.py::test_scatter_2d - RuntimeError: [insert_scratch_pad.cpp:initialize_sc...
FAILED tests/python/test_bls.py::test_gather_2d - RuntimeError: [insert_scratch_pad.cpp:initialize_scr...
FAILED tests/python/test_bls.py::test_scatter_1d - RuntimeError: [insert_scratch_pad.cpp:initialize_sc...
FAILED tests/python/test_bls_assume_in_range.py::test_gathering - RuntimeError: [insert_scratch_pad.cp...

File: python/taichi/lang/matrix.py [hunk 9]
```diff
@@ -1006,7 +1049,7 @@ def __repr__(self):
 
     @staticmethod
     @taichi_scope
-    def zero(dt, n, m=1):
+    def zero(dt, n, m=None):
```
[the remark this answers is not in the document]
Comment (squarefk) [reply python/taichi/lang/matrix.py:1048]: https://github.com/taichi-dev/taichi/pull/2637/commits/d6b503e74e2faaf064302a3ae92d7ea6c5b7ead4

File: python/taichi/lang/matrix.py (hunk 6 above)
Comment (squarefk) [reply python/taichi/lang/matrix.py]: https://github.com/taichi-dev/taichi/pull/2637/commits/0262df20e03d21b21d9b8f87d7c0ea6552578819

File: python/taichi/lang/impl.py (hunk 4 above)
Comment (k-ye) [reply python/taichi/lang/impl.py:21]: Thanks! `make_block_local` has a higher precedence, so let's throw an `InvalidOperationError` here,

File: python/taichi/lang/matrix.py [hunk 10]
```diff
@@ -85,11 +87,45 @@ def __init__(self,
                     if keep_raw:
                         mat = [list([x]) for x in n]
                     else:
-                        mat = [list([expr.Expr(x)]) for x in n]
+                        if not ti.is_extension_supported(
+                                ti.cfg.arch, ti.extension.dynamic_index
+                        ) or in_python_scope() or disable_local_tensor:
+                            mat = [list([expr.Expr(x)]) for x in n]
+                        else:
+                            dt = ti.default_cfg().default_ip if isinstance(
```
Comment (k-ye) [inline python/taichi/lang/matrix.py:95]: I believe it's possible to have `x = ti.Vector([a, b, c])`, where `a, b, c` themselves are expressions. In that case we need to use their types

File: taichi/ir/statements.h (hunk 3 above)
Comment (k-ye) [reply taichi/ir/statements.h:335]: yeah, something like that

**Review by strongoier — APPROVED** [review]

File: taichi/analysis/alias_analysis.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (xumingkuan) [reply taichi/analysis/alias_analysis.cpp:33]: IIUC, when `var1 != var2` but `origin1 == origin2`, we can write this?
```C++
auto diff = value_diff_ptr_index(var1->as<PtrOffsetStmt>()->offset, var2->as<PtrOffsetStmt>()->offset);
if (diff.is_diff_certain) {
  return diff.diff_range == 0 ? AliasResult::same : AliasResult::different;
} else {
  return AliasResult::uncertain;
}
```

File: taichi/analysis/alias_analysis.cpp [hunk 11]
```diff
@@ -13,8 +13,39 @@ AliasResult alias_analysis(Stmt *var1, Stmt *var2) {
...
+    if (origin1->is<AllocaStmt>() || origin2->is<AllocaStmt>())
+      return AliasResult::different;
+    TI_ASSERT(origin1->is<GlobalTemporaryStmt>() &&
+              origin2->is<GlobalTemporaryStmt>());
+    if (origin1->cast<GlobalTemporaryStmt>()->offset ==
+        origin2->cast<GlobalTemporaryStmt>()->offset) {
+      return AliasResult::uncertain;
```
Comment (xumingkuan) [inline taichi/analysis/alias_analysis.cpp:40]: Same as above -- maybe we want a function for that.

**Review by xumingkuan — COMMENTED** [review]
Great work! I think the alias analysis is a bit conservative in that it doesn't check the offset though.

File: taichi/analysis/alias_analysis.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (squarefk) [reply taichi/analysis/alias_analysis.cpp:33]: Good to know how to use ```value_diff_ptr_index```!

File: taichi/ir/statements.h (hunk 3 above)
Comment (squarefk) [reply taichi/ir/statements.h]: https://github.com/taichi-dev/taichi/pull/2637/commits/aee940856346702e2dafd47ebb1636afbae29801

**Comment by k-ye** [comment]
resolve conflicts?

File: taichi/analysis/alias_analysis.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (squarefk) [reply taichi/analysis/alias_analysis.cpp:33]: Let's do this in another PR.

File: taichi/analysis/alias_analysis.cpp (hunk 11 above)
Comment (squarefk) [reply taichi/analysis/alias_analysis.cpp:40]: Let's do this in another PR.

File: python/taichi/lang/impl.py (hunk 4 above)
Comment (squarefk) [reply python/taichi/lang/impl.py]: https://github.com/taichi-dev/taichi/pull/2637/commits/d2af38ee784bbe1624cd41fe476515f1e85c2c19

File: python/taichi/lang/matrix.py (hunk 10 above)
Comment (squarefk) [reply python/taichi/lang/matrix.py]: https://github.com/taichi-dev/taichi/pull/2637/commits/607d2f11532e1bac42a6f01d2ec1041c70ecff1a

**Pull request merged**